File: src/com/nerdscorner/mvp/mvp/FragmentMvpBuilder.java
package com.nerdscorner.mvp.mvp;

import com.intellij.openapi.vfs.VirtualFile;
import com.nerdscorner.mvp.mvp.busevents.fragment.FragmentComponent;
import com.nerdscorner.mvp.mvp.busevents.layout.LayoutComponent;
import com.nerdscorner.mvp.mvp.busevents.model.ModelComponent;
import com.nerdscorner.mvp.mvp.busevents.presenter.FragmentPresenterComponent;
import com.nerdscorner.mvp.mvp.busevents.view.FragmentViewComponent;
import com.nerdscorner.mvp.utils.GradleUtils;


import static com.nerdscorner.mvp.utils.GradleUtils.MVP_LIB_EVENTS_DEPENDENCY;

public class FragmentMvpBuilder extends MvpBuilder {

    public FragmentMvpBuilder(boolean shouldIncludeLibraryDependency, boolean isJava, boolean shouldCreateWiring) {
        super(shouldIncludeLibraryDependency, isJava, shouldCreateWiring);
    }

    @Override
    public boolean build(VirtualFile rootFolder, String fullPath, String packageName, String screenName) {
        FragmentComponent fragmentComponent = new FragmentComponent(fullPath, packageName, screenName, shouldCreateWiring);
        ModelComponent modelComponent = new ModelComponent(fullPath, packageName, screenName, shouldCreateWiring);
        FragmentViewComponent fragmentViewComponent = new FragmentViewComponent(fullPath, packageName, screenName, shouldCreateWiring);
        FragmentPresenterComponent fragmentPresenterComponent = new FragmentPresenterComponent(fullPath, packageName, screenName, shouldCreateWiring);
        LayoutComponent layoutComponent = new LayoutComponent(fullPath, packageName, screenName, false);
        boolean success = fragmentComponent.build(isJava);
        success = success && modelComponent.build(isJava)
                && fragmentViewComponent.build(isJava)
                && fragmentPresenterComponent.build(isJava)
                && layoutComponent.build();
        success = updateGradle(rootFolder, success);
        if (!success) {
            rollback(rootFolder, fragmentComponent, modelComponent, fragmentViewComponent, fragmentPresenterComponent, layoutComponent);
        }
        return success;
    }

    private boolean updateGradle(VirtualFile rootFolder, boolean success) {
        if (shouldIncludeLibraryDependency) {
            savedGradleFile = GradleUtils.getGradleFileContent(rootFolder);
            success = success && GradleUtils.addDependency(rootFolder, MVP_LIB_EVENTS_DEPENDENCY);
        }
        return success;
    }

    private void rollback(VirtualFile rootFolder, BaseComponent activityComponent,
                          BaseComponent modelComponent, BaseComponent viewComponent, BaseComponent presenterComponent,
                          LayoutComponent layoutComponent) {
        activityComponent.rollback();
        modelComponent.rollback();
        viewComponent.rollback();
        presenterComponent.rollback();
        layoutComponent.rollback();
        GradleUtils.restoreGradleFile(savedGradleFile, rootFolder);
    }
}

File: src/com/nerdscorner/mvp/mvp/FragmentMvpBuilder.kt
package com.nerdscorner.mvp.mvp

import com.intellij.openapi.vfs.VirtualFile
import com.nerdscorner.mvp.domain.ExecutionResult
import com.nerdscorner.mvp.mvp.busevents.fragment.FragmentComponent
import com.nerdscorner.mvp.mvp.busevents.layout.LayoutComponent
import com.nerdscorner.mvp.mvp.busevents.model.ModelComponent
import com.nerdscorner.mvp.mvp.busevents.presenter.FragmentPresenterComponent
import com.nerdscorner.mvp.mvp.busevents.view.FragmentViewComponent
import com.nerdscorner.mvp.utils.GradleUtils

class FragmentMvpBuilder(shouldIncludeLibraryDependency: Boolean, isJava: Boolean, shouldCreateWiring: Boolean) :
        MvpBuilder(shouldIncludeLibraryDependency, isJava, shouldCreateWiring) {
    private var fragmentFileCreated = ExecutionResult.EMPTY

    override fun build(rootFolder: VirtualFile, fullPath: String, packageName: String, screenName: String): ExecutionResult {
        val fragmentComponent = FragmentComponent(fullPath, packageName, screenName, shouldCreateWiring)
        val modelComponent = ModelComponent(fullPath, packageName, screenName, shouldCreateWiring)
        val viewComponent = FragmentViewComponent(fullPath, packageName, screenName, shouldCreateWiring)
        val presenterComponent = FragmentPresenterComponent(fullPath, packageName, screenName, shouldCreateWiring)
        val layoutComponent = LayoutComponent(fullPath, packageName, screenName, false)

        fragmentFileCreated = fragmentComponent.build(isJava)
        modelFileCreated = modelComponent.build(isJava)
        viewFileCreated = viewComponent.build(isJava)
        presenterFileCreated = presenterComponent.build(isJava)
        layoutFileCreated = layoutComponent.build()

        val executionResult = fragmentFileCreated +
                              modelFileCreated +
                              viewFileCreated +
                              presenterFileCreated +
                              layoutFileCreated +
                              updateGradleFile(rootFolder)
        if (executionResult.successful.not()) {
            rollback(rootFolder, fragmentComponent, modelComponent, viewComponent, presenterComponent, layoutComponent)
        }
        return executionResult
    }

    private fun rollback(rootFolder: VirtualFile,
                         fragmentComponent: BaseComponent,
                         modelComponent: BaseComponent,
                         viewComponent: BaseComponent,
                         presenterComponent: BaseComponent,
                         layoutComponent: LayoutComponent) {
        if (fragmentFileCreated.successful) {
            fragmentComponent.rollback()
        }
        if (modelFileCreated.successful) {
            modelComponent.rollback()
        }
        if (viewFileCreated.successful) {
            viewComponent.rollback()
        }
        if (presenterFileCreated.successful) {
            presenterComponent.rollback()
        }
        if (layoutFileCreated.successful) {
            layoutComponent.rollback()
        }
        GradleUtils.restoreGradleFile(savedGradleFile, rootFolder)
    }
}

